drivers/ra/fsp/src/r_usb_device/r_usb_device.c

```diff
@@ -0,0 +1,1478 @@
+
+/**
+ * @brief Start USB module, configure correct operate mode and default pipe0
+ *
+ * @param p_api_ctrl    [in]
+ * @param p_cfg         [in]
+ *
+ * @retval FSP_SUCCESS          on success
+ * @retval FSP_ERR_UNSUPPORTED  if USB module just support FS only
+ */
+fsp_err_t R_USBD_Open (usbd_ctrl_t * const p_api_ctrl, usbd_cfg_t const * const p_cfg)
+{
+    usbd_instance_ctrl_t * p_ctrl = (usbd_instance_ctrl_t *) p_api_ctrl;
+
+#if USBD_CFG_PARAM_CHECKING_ENABLE
+    FSP_ASSERT(p_ctrl);
+    FSP_ASSERT(p_cfg);
+    FSP_ERROR_RETURN(0 == p_ctrl->open, FSP_ERR_ALREADY_OPEN);
+
+    /* Make sure this channel exists. */
+    FSP_ERROR_RETURN(USB_MODULE_COUNT > p_cfg->module_number, FSP_ERR_IP_CHANNEL_NOT_PRESENT);
+
+    /* USBFS module is not currently supported. */
+    FSP_ERROR_RETURN(USB_HS_MODULE_NUMBER_START >= p_cfg->module_number, FSP_ERR_UNSUPPORTED);
+
+    /* USBHS module only support FS/HS operation */
+    FSP_ERROR_RETURN((USBD_SPEED_FS == p_cfg->usb_speed) || (USBD_SPEED_HS == p_cfg->usb_speed),
+                     FSP_ERR_INVALID_ARGUMENT);
+#endif
+
+    p_ctrl->p_reg             = (void *) R_USB_HS0_BASE;
+    p_ctrl->p_cfg             = p_cfg;
+    p_ctrl->p_callback        = p_cfg->p_callback;
+    p_ctrl->p_context         = p_cfg->p_context;
+    p_ctrl->p_callback_memory = NULL;
+
+    /* clear pipe config */
+    memset(&g_pipe_cfg[p_cfg->module_number], 0, sizeof(usb_pipe_cfg_t));
+
+    /* module start */
+    R_BSP_MODULE_START(FSP_IP_USBHS, 0);
+
+    /* physical layer init */
+    physical_init(p_ctrl, p_cfg);
+
+    /* setup default control pipe */
+    control_pipe_setup(p_api_ctrl);
+
+    /* enable interrupt */
+    R_BSP_IrqCfgEnable(p_cfg->hs_irq, p_cfg->hsipl, p_ctrl);
+
+    p_ctrl->open = USBD_OPEN;
+
+    return FSP_SUCCESS;
+}
+
+/**
+ * @brief Shutdown USB device
+ *
+ * @param p_api_ctrl
+ * @param p_cfg
+ *
+ * @retval FSP_SUCCESS on success
+ */
+fsp_err_t R_USBD_Close (usbd_ctrl_t * const p_api_ctrl)
+{
+    usbd_instance_ctrl_t * p_ctrl = (usbd_instance_ctrl_t *) p_api_ctrl;
+
+#if USBD_CFG_PARAM_CHECKING_ENABLE
+    FSP_ASSERT(p_ctrl);
+    FSP_ERROR_RETURN(0 != p_ctrl->open, FSP_ERR_NOT_OPEN);
+#endif
+
+    /* disable interrupt */
+    R_BSP_IrqDisable(p_ctrl->p_cfg->hs_irq);
+
+    usb_module_register_clear(p_ctrl->p_cfg->module_number);
+
+    /* module stop */
+    R_BSP_MODULE_STOP(FSP_IP_USBHS, 0);
+
+    p_ctrl->open = 0;
+
+    return FSP_SUCCESS;
+}
+
+/**
+ * @brief Connect USB D+/D- pin to the USB data bus
+ *
+ * @param p_api_ctrl
+ *
+ * @retval FSP_SUCCESS on success
+ */
+fsp_err_t R_USBD_Connect (usbd_ctrl_t * const p_api_ctrl)
+{
+#if USBD_CFG_PARAM_CHECKING_ENABLE
+    usbd_instance_ctrl_t * p_ctrl = (usbd_instance_ctrl_t *) p_api_ctrl;
+
+    FSP_ASSERT(p_ctrl);
+    FSP_ERROR_RETURN(0 != p_ctrl->open, FSP_ERR_NOT_OPEN);
+#else
+    FSP_PARAMETER_NOT_USED(p_api_ctrl);
+#endif
+
+    R_USB_HS0->SYSCFG_b.CNEN  = 1;
+    R_USB_HS0->SYSCFG_b.DPRPU = 1;
```

Comment:
Using a p_reg pointer is not really needed, because on all devices there is only 1 channel for USBHS module, even on upcoming RA8x2 devices. Besides that, current implementation of FSP USB also use the IODefine from BSP (R_USB_HS0, R_USB_FS0) directly, instead of using a p_reg pointer